pkg/controller/virtualmachineimport/virtualmachineimport_controller.go

```diff
@@ -635,54 +641,57 @@ func (r *ReconcileVirtualMachineImport) createDataVolumes(provider provider.Prov
-		if err := controllerutil.SetControllerReference(instance, &dv, r.scheme); err != nil {
-			return err
-		}
 
-		// Set tracking label
-		setTrackerLabel(dv.ObjectMeta, instance)
+	// Set controller owner reference:
+	if err := controllerutil.SetControllerReference(instance, &dv, r.scheme); err != nil {
+		return err
+	}
 
-		err = r.client.Create(context.TODO(), &dv)
-		if err != nil {
-			message := fmt.Sprintf("Data volume %s/%s creation failed: %s", dv.Namespace, dv.Name, err)
-			// Update event:
-			r.recorder.Event(instance, corev1.EventTypeWarning, EventDVCreationFailed, message)
-
-			// Update condition to failed:
-			succeededCond := conditions.NewSucceededCondition(string(v2vv1alpha1.DataVolumeCreationFailed), message, corev1.ConditionFalse)
-			processingCond.Status = corev1.ConditionFalse
-			processingFailedReason := string(v2vv1alpha1.ProcessingFailed)
-			processingCond.Reason = &processingFailedReason
-			if err = r.upsertStatusConditions(instanceNamespacedName, processingCond, succeededCond); err != nil {
-				return err
-			}
+	// Set tracking label
+	setTrackerLabel(dv.ObjectMeta, instance)
 
-			// Cleanup
-			if err = r.afterFailure(instanceNamespacedName, provider); err != nil {
-				return err
-			}
+	err = r.client.Create(context.TODO(), &dv)
+	if err != nil {
+		message := fmt.Sprintf("Data volume %s/%s creation failed: %s", dv.Namespace, dv.Name, err)
+		// Update event:
+		r.recorder.Event(instance, corev1.EventTypeWarning, EventDVCreationFailed, message)
+
+		// Update condition to failed:
+		succeededCond := conditions.NewSucceededCondition(string(v2vv1alpha1.DataVolumeCreationFailed), message, corev1.ConditionFalse)
+		processingCond.Status = corev1.ConditionFalse
+		processingFailedReason := string(v2vv1alpha1.ProcessingFailed)
+		processingCond.Reason = &processingFailedReason
+		if err = r.upsertStatusConditions(instanceNamespacedName, processingCond, succeededCond); err != nil {
 			return err
 		}
 
-		// Set VM as owner reference:
-		if err := r.ownerreferencesmgr.AddOwnerReference(vmDef, &dv); err != nil {
+		// Cleanup
+		if err = r.afterFailure(instanceNamespacedName, provider); err != nil {
 			return err
 		}
+		return err
 	}
 
-	// Emit event that DVs import is in progress:
-	r.recorder.Eventf(instance, corev1.EventTypeNormal, EventImportInProgress, "Import of Virtual Machine %s/%s disks are in progress", vmName.Namespace, vmName.Name)
+	// Set VM as owner reference:
+	if err := r.ownerreferencesmgr.AddOwnerReference(vmDef, &dv); err != nil {
+		return err
+	}
+
+	// Emit event that DV import is in progress:
+	r.recorder.Eventf(
+		instance,
+		corev1.EventTypeNormal,
+		EventImportInProgress,
+		"Import of Virtual Machine %s/%s disk %s in progress", vmName.Namespace, vmName.Name, dv.Name,
+	)
 
 	// Update datavolume in VM import CR status:
-	if err = r.updateDVs(instanceNamespacedName, dvs); err != nil {
+	if err = r.updateDVs(instanceNamespacedName, dv); err != nil {
```

Comment:
@machacekondra do you see any option for this method or `updateVMSpecDataVolumes` be called more than once for the created DV ? Any probability of `updateDVs` to succeed and `updateVMSpecDataVolumes` to fail ?

I'd like to make sure this method will not be called more than once per each DV and to make sure the list of reported DV for the vmimport CR's status is updated correctly.

Reply:
Well, I can think of any case, when it can happen, but I will improve those two calls, so it's more reliable, I've opened an [issue](https://github.com/kubevirt/vm-import-operator/issues/290), and will fix in separate PR.